src/LibraryUtilities.ts

```diff
@@ -566,48 +281,194 @@ export function updateSections(oldLayoutSpecs: any, newLayoutSpecs: any): void {
+    let nodeInculde = false;
+
+    for (let i = 0; i < includeInfo.length && t < typeListNodes.length; i++) {
+        nodeMatch = false;
+        nodeInculde = false;
+        let fullyQualifiedNameParts = typeListNodes[t].fullyQualifiedName.split(".");
+
+        if (includeInfo[i].path.indexOf(prefix) == -1) {
+            let includeParts = includeInfo[i].path.split(".");
+            let compareResult = compareParts(fullyQualifiedNameParts, includeParts);
+
+            if (compareResult == 0) {
+                nodeMatch = true;
+            } else if (compareResult > 0) {
+                nodeInculde = true;
+                let newName = fullyQualifiedNameParts.slice(-compareResult - 1).join('.');
+                buildLibraryItemsFromName(typeListNodes[t], includeInfo[i].parentItem, newName, includeInfo[i].iconUrl);
+            } else if (typeListNodes[t].fullyQualifiedName.localeCompare(includeInfo[i].path) < 0) {
+                i--;
+                t++;
+            }
 
-    // Search for the nodes that are not displayed in library view.
-    let unprocessedNodes: TypeListNode[] = [];
-    let sectionData = convertSectionToItemData(section);
+        } else if (typeListNodes[t].fullyQualifiedName.startsWith(includeInfo[i].path)) {
+            nodeInculde = true;
+            let prefixIndex = typeListNodes[t].fullyQualifiedName.indexOf(prefix);
+            let newName = typeListNodes[t].fullyQualifiedName.substring(prefixIndex + prefix.length);
+            buildLibraryItemsFromName(typeListNodes[t], includeInfo[i].parentItem, newName);
+        }
 
-    _.each(allNodes, function (node) {
-        if (!node.processed) {
-            console.warn("Item filtered out from library view: " + node.contextData);
-            unprocessedNodes.push(node);
+        if (nodeMatch) {
+            let item = new ItemData(fullyQualifiedNameParts[fullyQualifiedNameParts.length - 1]);
+            item.constructFromTypeListNode(typeListNodes[t]);
+            includeInfo[i].parentItem.appendChild(item);
         }
-    });
 
-    _.each(unprocessedNodes, function (node) {
-        buildLibraryItemsFromName(node, sectionData)
-    });
+        if (nodeMatch || nodeInculde) {
+            typeListNodes[t].processed = true;
+            i--;
+            t++;
+        }
+    }
+}
+
+export function buildLibrarySectionsFromLayoutSpecs(loadedTypes: any, layoutSpecs: any, defaultSectionStr: string, miscSectionStr: string): ItemData[] {
+
+    let typeListNodes: TypeListNode[] = [];
+    let sectionElements: LayoutElement[] = [];
+
+    // Converting raw data to strongly typed data.
+    for (let loadedType of loadedTypes.loadedTypes) {
+        typeListNodes.push(new TypeListNode(loadedType));
+    }
+
+    for (let section of layoutSpecs.sections) {
+        sectionElements.push(new LayoutElement(section));
+    }
+
+    let sections = convertLayoutElementToItemData(sectionElements);
+    let includeInfo = getIncludeInfo(sectionElements);
```

Comment:
includeInfo sounds like a single value, you can change the name to includeInfoList.